Complexity_classes_def/P.thy

```diff
@@ -401,10 +401,11 @@ definition IMP_SAT :: "nat set" where
 "IMP_SAT == encode_sat ` {n. sat n}"
 
 lemma main_lemma : 
-  fixes c p
-  assumes "poly p"
-  assumes "\<forall>s. \<exists>t. Ex (big_step_t (c, s) t) \<and> t \<le> p (length (s ''input''))"
-  assumes " \<forall>x z. \<exists>r. \<forall>s. s ''input'' = x \<and> s ''certificate'' = z \<longrightarrow>
+  fixes c pt p_cer
```

Comment:
New main-lemma:
Before reading it , you shpould know I added the following assumption to the definition of a polynomial verifier: 
"if a valid certificate exists, then there's a  valid certificate that is polynomially bounded in terms of the input length" Let's call this what "a bound on the certificate length" means.
Now the lemma 
fixing:
 - a verifier code c: 
- a polynomial time bound of the verifier 
- a polynomial certificate bound of the verifier

we assume the things that I thought it might be helpful to assume, which are convenient with our defintions :
- the verifier's code always terminates on a polynomial time in terms of input length. (and only the input ,not the certificate)
I just assumed that all verifiers would terminate immediately if they realised the certificate is too big. In a SAT formula that could mean a certificate that is bigger than the formula and so on.  Due the certificate bound, it shouldn't actually be a problem for stating cook-levin because a polynomial time bound leads to the other.

I don't claim this (a time bound only in terms of the input length ) is really necessary. 
Pros: 
* It does makes things easier as it does provide a unified simple definition of time bound for both deciders and verifiers. Thus made it easier to state cook-levin. 
Cons: 
* Proving a problem is NP (without reduction /for example: SAT) could be a bit harder since the time bound is a bit harsher. the verifier should get sure the certificate isn't too long as said.

I don't know what is actually more helpful for the reduction, or is it even related? Anyway we could change it in the definition of NP and in the main-lemma if necessary

also to be assumed:
-   a result of a verifier is only dependent on the input and the certificate. (Pretty obvious)
The assumptions also give a polynomial bound on the certificate without saying that "a certificate is bounded" , because I thought it is irrelevant to the work of the reduction. The reduction should just use this bound and tell whether a certificate exists under this bound. It doesn't need for that to know that the certificate is inherently bounded.

To be shown through the work that is to be done: 
- There exists a reduction that takes the input of  the verifier, gives back a  SAT formula that is satisfiable iff there exists a certificate under the given bound that would be accepted with the given input. 
- The reduction should have some polynomial bound in time in terms of length of the input and give a formula that is polynomially bounded in length.
- The reduction shouldn't user the register certificate (Well there's an infinity of registers, so ..)

If some assumptions are irrelevant to the work, they could be easily removed, if others are needed, please let me konw.